siliconcompiler/tools/vpr/_json_constraint.py

```diff
@@ -0,0 +1,98 @@
+import argparse
+import json
+
+from siliconcompiler.tools.vpr._xml_constraint import generate_vpr_constraints_xml
+from siliconcompiler.tools.vpr._xml_constraint import write_vpr_constraints_xml_file
+
+
+def main():
+
+    option_parser = argparse.ArgumentParser()
+    option_parser.add_argument("-constraints_map",
+                               help="architecture-specific mapping of constraint pins "
+                                    "to FPGA core pins")
+    option_parser.add_argument("json_constraints",
+                               help="architecture-independent constraints file")
+    option_parser.add_argument("constraints_file_out",
+                               help="constraints XML file name")
+
+    options = option_parser.parse_args()
+
+    json_constraints_file = options.json_constraints
+    constraints_map_file = options.constraints_map
+    constraints_file_out = options.constraints_file_out
+
+    json_generic_constraints = load_json_constraints(json_constraints_file)
+    if (constraints_map_file):
+        constraints_map = load_constraints_map(constraints_map_file)
+    else:
+        constraints_map = {}
+
+    mapped_constraints = map_constraints(json_generic_constraints,
+                                         constraints_map)
+
+    constraints_xml = generate_vpr_constraints_xml(mapped_constraints)
+    write_vpr_constraints_xml_file(constraints_xml, constraints_file_out)
+
+
+def load_json_constraints(json_constraints_file):
+
+    json_generic_constraints = {}
+    with (open(json_constraints_file, "r")) as json_constraints_data:
+        json_generic_constraints = json.loads(json_constraints_data.read())
+
+    return json_generic_constraints
+
+
+def load_constraints_map(constraints_map_file):
+
+    constraints_map = {}
+    with (open(constraints_map_file, "r")) as constraints_map_data:
+        constraints_map = json.loads(constraints_map_data.read())
+
+    return constraints_map
+
+
+def map_constraints(json_generic_constraints,
```

Comment:
is there any error checking available? ie. if I set a pin to a non-existent location?

Reply:
For checking that locations are valid, pretty robust error checking should be possible if we add a VPR XML file parser.  Need to decide a) if you want to sign up for supporting said parser; b) on what timetable to do the development.  More rudimentary error checking is possible if FPGA array size is supplied, as you can do out of bounds checking.  If we check for in-bounds errors in assignment without XML support the checking is likely to be quite brittle.  

Also possible to add error checking for pin name correctness by adding a BLIF parser; then you could report whether the JSON file has a bad pin name.  Similarly, and without any extra parsers, we can check that the user has supplied port names that are in the constraints map JSON file for mapping to the FPGA architecture.